### PR TITLE
Add cross-attention to output hypotheses

### PR DESCRIPTION
**Pull request by mgaido91**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

The PR adds the encoder-decoder cross-attention to the output hypotheses returned by ASR models.

**Collection**: ASR

# Changelog

- Returns the cross-attention scores in the output of the greedy generator
- Returns the cross-attention scores in the output of the beam search generator

# Usage

- You can potentially add a usage example below

```python
from nemo.collections.asr.models import ASRModel
from nemo.collections.asr.models.aed_multitask_models import MultiTaskTranscriptionConfig
from nemo.collections.asr.parts.submodules.multitask_decoding import MultiTaskDecodingConfig

model = ASRModel.from_pretrained(model_name="nvidia/canary-1b-v2")
decoding_config = MultiTaskDecodingConfig()
decoding_config.return_xattn_scores = True
model.change_decoding_strategy(decoding_config)
config = MultiTaskTranscriptionConfig(
    batch_size=4,
    return_hypotheses=True,
    num_workers=0,
    verbose=False,
    prompt={'source_lang': 'en', 'target_lang': 'en'},
    enable_chunking=False
)
output = model.transcribe("/Users/mgaido/Downloads/vp-test/aa.wav", override_config=config)
assert output[0].xatt_scores is not None
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

@nithinraok @andrusenkoau 

# Additional Information

- Related to #14961 .

### CODE REVIEW

File: nemo/collections/asr/parts/utils/rnnt_utils.py [hunk 1]
```diff
@@ -87,6 +87,9 @@ class Hypothesis:
     last_token (Optional): A token or batch of tokens which was predicted in the last step.
 
     last_frame (Optional): Index of the last decoding step hypothesis was updated including blank token prediction.
+
+    xatt_scores (Optional): List of cross-attention scores for each decoder layer. Each element of the list
```
Comment (nithinraok) [inline nemo/collections/asr/parts/utils/rnnt_utils.py]: Shouldn;t shape is List[`BxHxT1xT2`] . Also best to add: this is used with AED models

File: nemo/collections/asr/modules/transformer/transformer_generators.py [hunk 2]
```diff
@@ -234,14 +235,19 @@ def _forward(
...
                 input_ids,
                 encoder_hidden_states,
                 encoder_input_mask,
                 decoder_mems_list,
                 i,
                 return_scores=return_beam_scores,
             )
+            if xatt_scores_list is not None:
+                for layer in range(len(xatt_scores_list)):
+                    xatt_scores_list[layer] = torch.cat((xatt_scores_list[layer], new_xatt_scores_list[layer]), dim=2)
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer/transformer_generators.py]: what about condition when `new_xattn_scores_list` is None? cat would fail

File: nemo/collections/asr/modules/transformer/transformer_generators.py
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer/transformer_generators.py:374]: could you update here as well and also include in returns tuple

File: nemo/collections/asr/modules/transformer/transformer_generators.py [hunk 3]
```diff
@@ -478,6 +488,21 @@ def _forward(
             prefixes_ids = indices_i.unsqueeze(2).repeat(1, 1, p_len)
             prefixes = prefixes.gather(1, prefixes_ids).view(-1, p_len)
 
+            # select xatt scores corresponding to chosen hypotheses
+            if next_xatt_scores_list is not None:
+                num_heads = xatt_scores_list[0].shape[1]
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer/transformer_generators.py:502]: check for `xatt_scores_list` if None

File: nemo/collections/asr/modules/transformer/transformer_generators.py [hunk 4]
```diff
@@ -501,13 +526,26 @@ def _forward(
...
+                    .view(-1, self.beam_size, num_heads, tgt_len, src_len)
+                    .gather(1, xatt_best_guesses)
+                    .squeeze(1)
+                )
 
         if return_beam_scores:
-            return prefixes, scores * len_penalties, tgt
+            return prefixes, scores * len_penalties, tgt, xatt_scores_list
         else:
             return tgt
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer/transformer_generators.py:559]: we might also return `xatt_scores_list` here as `return_beam_scores` is independent of return xattn_scores.

File: nemo/collections/asr/modules/transformer/transformer_generators.py (hunk 2 above)
Comment (mgaido91) [reply nemo/collections/asr/modules/transformer/transformer_generators.py]: if `new_xattn_scores_list` is `None`, `xatt_scores_list` will stay `None`, so we never enter in this if

File: nemo/collections/asr/modules/transformer/transformer_generators.py (hunk 3 above)
Comment (mgaido91) [reply nemo/collections/asr/modules/transformer/transformer_generators.py:502]: if `next_xatt_scores_list` is not None, then `xatt_scores_list` cannot be None. `_one_step_forward` either always returns None or always returns a not-None value

File: nemo/collections/asr/parts/utils/rnnt_utils.py (hunk 1 above)
Comment (mgaido91) [reply nemo/collections/asr/parts/utils/rnnt_utils.py]: this is for a single hypothesis, so there is no `B`... So this would be `List[HxT1xT2]`. If you prefer I can rename `HxUxT` to `HxT1xT2`. I will also add the indication about AED models.

File: nemo/collections/asr/modules/transformer/transformer_generators.py (hunk 4 above)
Comment (mgaido91) [reply nemo/collections/asr/modules/transformer/transformer_generators.py:559]: maybe I am missing something, but I see this being used with `return_beam_scores=True`, while `return_beam_scores` is set to False only in the validation step in `EncDecTransfModelBPE`, where there is no point in returning the attention scores. So I thought it would be better to avoid an API change for no reason. I can implement that, but just wanted to double check whether we really want this.

**Comment by mgaido91** [comment]
@nithinraok thanks for your review. I do have a question, though. You said to add a param in `MultiTaskDecodingConfig`, however IIUC we should rather focus on `TranscribeConfig`, which is what is set through the `transcribe` API. Changing the `MultiTaskDecodingConfig ` might be tricky for a user. Also, there are some config tricks required in the `TrascribeConfig` (i.e. `enable_chunking=False`) to enable getting the attention. So if we add a param in `MultiTaskDecodingConfig`, it might be quite complex for a user to get the attention, as it would require changing two different configs. WDYT?

File: nemo/collections/asr/modules/transformer/transformer_generators.py (hunk 2 above)
Comment (nithinraok) [reply nemo/collections/asr/modules/transformer/transformer_generators.py]: I meant to ask for each step, but probably its fine.

**Comment by nithinraok** [comment]
> @nithinraok thanks for your review. I do have a question, though. You said to add a param in `MultiTaskDecodingConfig`, however IIUC we should rather focus on `TranscribeConfig`, which is what is set through the `transcribe` API. Changing the `MultiTaskDecodingConfig ` might be tricky for a user. Also, there are some config tricks required in the `TrascribeConfig` (i.e. `enable_chunking=False`) to enable getting the attention. So if we add a param in `MultiTaskDecodingConfig`, it might be quite complex for a user to get the attention, as it would require changing two different configs. WDYT?

Very good point. 

Well, first, the option need to be passed to `MultiTaskDecodingConfig` for sure. To make it independent of the behavior of transcribe inference or not. But however, I get your point. 
Two options:
* If the option needs to be through `TranscribeConfig` then we need to run `model.change_decoding_strategy(updated_cfg)` to change the default decoding behavior (which is return xattn_scores=None) inside .transcribe() each time. Which I don;t think is a good idea. 
* 2nd options is we create decoding cfg. like:
```python
multitask_decoding = MultiTaskDecodingConfig()
multitask_decoding.strategy = "greedy"
multitask_decoding.return_xattn_scores = True
```
and call
```python
asr_model.change_decoding_strategy(multitask_decoding)
```
before performing .transcribe()

Lets go with later option for now. I will keep thinking about this as this needs to be changed across. Are there any other options like these you are interested on to be changed through .transcribe()?

File: nemo/collections/asr/modules/transformer/transformer_generators.py (hunk 2 above)
Comment (mgaido91) [reply nemo/collections/asr/modules/transformer/transformer_generators.py]: I got it but if `xatt_scores_list` is not None, then `new_xattn_scores_list` must be not None, otherwise there is something weird happening (ie. attention scores are returned for some tokens and for some others are None in the same generation). I'd rather add an assert on `new_xattn_scores_list`, WDYT?

**Comment by mgaido91** [comment]
> Lets go with later option for now. I will keep thinking about this as this needs to be changed across. Are there any other options like these you are interested on to be changed through .transcribe()?

Ok, I will work on this in the next days. Maybe it would be worth adding some checks and logs to guide the user though. I will try to come up with a proposal for that while working on this. Thanks.

File: tests/collections/asr/decoding/test_multi_task_decoding.py [hunk 5]
```diff
@@ -103,11 +111,11 @@ def test_greedy_decoding(inputs, nnet, deterministic_rng, with_confidence):
 
 
 def test_temperature_sampling_decoding(inputs, nnet):
-    gen = GreedySequenceGenerator(*nnet, temperature=10.0, n_samples=2)
+    gen = GreedySequenceGenerator(*nnet, return_xattn_scores=True, temperature=10.0, n_samples=2)
```
Comment (andrusenkoau) [inline tests/collections/asr/decoding/test_multi_task_decoding.py]: Could you add the check for both `return_xattn_scores` options (as above) here?

**Review by andrusenkoau — COMMENTED** [review]
Hi @mgaido91, thank you for the great work! I have almost no questions. Let's wait your final changes with decoding config.

**Comment by mgaido91** [comment]
Thanks @andrusenkoau !

> Let's wait your final changes with decoding config.

I already made them. You can find the `return_xattn_scores` in the decoding config. I also added a warning message to instruct the users. Is there something missing that you have in mind?


PS I do not see why `[CICD NeMo / Nemo_CICD_Test (pull_request)](https://github.com/NVIDIA-NeMo/NeMo/actions/runs/20991982507/job/60354816538?pr=15229)` is failing, I cannot see meaningful logs, maybe I just do not know where to look at as I am not that familiar with NeMo CI. It is also weird as it passed before I updated the branch with the master. May you point me to the logs?

File: tests/collections/asr/decoding/test_multi_task_decoding.py (hunk 5 above)
Comment (mgaido91) [reply tests/collections/asr/decoding/test_multi_task_decoding.py]: yes, sure, I did not do it to minimize the CI cost, I am updating it, thanks!

**Comment by mgaido91** [comment]
@nithinraok I think I addressed your comments, may you please take another look at this?

File: nemo/collections/asr/modules/transformer/transformer_generators.py
Comment (nithinraok) [reply nemo/collections/asr/modules/transformer/transformer_generators.py:374]: @mgaido91 could you update this one as well

**Review by nithinraok — APPROVED** [review]
Thanks Marco. Minor comment which you might have missed earlier. LGTM otherwise. Thanks for the PR, great work!

**Comment by mgaido91** [comment]
@nithinraok @andrusenkoau thank you for your guidance and reviews!
